File: src/24. Swap Nodes in Pairs/24.py
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None


class Solution:
    def swapPairs(self, head: ListNode) -> ListNode:
        # If the list has no node or has only one node left.
        if not head or not head.next:
            return head
        # Nodes to be swapped
        first_node = head
        second_node = head.next
        # Swapping
        first_node.next = self.swapPairs(second_node.next)
        second_node.next = first_node
        # Now the head is the second node
        return second_node

    def swapPairs1(self, head: ListNode) -> ListNode:
        pre, pre.next = self, head
        while pre.next and pre.next.next:
            a = pre.next
            b = a.next
            pre.next, b.next, a.next = b, a, b.next
            pre = a
        return self.next


def createList(nums):
    head = ListNode(0)
    cur = head
    for num in nums:
        cur.next = ListNode(num)
        cur = cur.next
    return head


def printList(head):
    cur = head
    while cur != None:
        print(cur.val, '->', end='')
        cur = cur.next
    print('NULL')


if __name__ == "__main__":
    nums = [1, 2, 3, 4]
    printList(Solution().swapPairs(createList(nums)))
    printList(Solution().swapPairs1(createList(nums)))

File: src/24. Swap Nodes in Pairs/24.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
 func swapPairs(head *ListNode) *ListNode {
	if head == nil || head.Next == nil {
		return head
	}
	result := head.Next
	head.Next = swapPairs(head.Next.Next)
	result.Next = head
	return result
}

 func swapPairs(head *ListNode) *ListNode {
    prev := new(ListNode)
    prev.Next = head
    t := prev
    for prev.Next != nil && prev.Next.Next != nil {
        cur := prev.Next
        ne  := cur.Next
        prev.Next,ne.Next,cur.Next = ne,cur,ne.Next
        prev = cur
    }
    return t.Next
}